aten/src/ATen/native/TensorShape.cpp

```diff
@@ -392,6 +390,86 @@ Tensor cat(TensorList tensors, int64_t dim) {
   return result;
 }
 
+Tensor block_diag(TensorList tensors) {
+  Tensor result;
+  if (tensors.size() == 0) {
+    result = at::empty({});
+    return result;
+  }
+
+  // Use at::result_type to find the right tensor to copy options from, so
+  // that the output data type is correct
+  const Tensor* options_tensor = &tensors[0];
+  ScalarType output_scalar_type = options_tensor->scalar_type();
+  const Device& device = tensors[0].device();
+
+  for (size_t tensor_idx = 1; tensor_idx < tensors.size(); tensor_idx++) {
+    const Tensor* other_tensor = &tensors[tensor_idx];
+
+    TORCH_CHECK(
+      other_tensor->device() == device,
+      "torch.block_diag: input tensors must all be on the same device"
+    );
+
+    ScalarType scalar_type = at::result_type(*options_tensor, *other_tensor);
+
+    if (scalar_type != output_scalar_type) {
+      output_scalar_type = scalar_type;
+      options_tensor = other_tensor;
+    }
+  }
+
+  int64_t result_dim0 = 0;
+  int64_t result_dim1 = 0;
+  std::vector<Tensor> tensors_2D(tensors.size());
+
+  // Sum the dimensions of the tensors, check tensor sizes,
+  // and expand all 0-D and 1-D tensors so that everything
+  // is 2-D
+  for (size_t tensor_idx = 0; tensor_idx < tensors.size(); tensor_idx++) {
+    const Tensor& tensor = tensors[tensor_idx];
+    int64_t ndims = tensor.dim();
+    TORCH_CHECK(ndims <= 2, "torch.block_diag: input tensors must have 2 or fewer dimensions");
+
+    int64_t dim0 = 1;
+    int64_t dim1 = 1;
+
+    if (ndims == 2) {
+      dim0 = tensor.size(0);
+      dim1 = tensor.size(1);
+      tensors_2D[tensor_idx] = tensor;
+    } else if (ndims == 1) {
+      // Switching dim 0 to dim 1 is intentional
+      dim1 = tensor.size(0);
+      tensors_2D[tensor_idx] = tensor.expand({dim0, dim1});
+    } else {
+      tensors_2D[tensor_idx] = tensor.expand({dim0, dim1});
+    }
+    result_dim0 += dim0;
+    result_dim1 += dim1;
+  }
+
+  result = at::zeros({result_dim0, result_dim1}, options_tensor->options());
+
+  int64_t cur_dim0 = 0;
+  int64_t cur_dim1 = 0;
+
+  // Copy each tensor into the appropriate location in the result matrix
+  std::vector<Tensor>::const_iterator iter;
+  for (iter = tensors_2D.begin(); iter != tensors_2D.end(); iter++) {
```

Comment:
I think I was wrong about my initial suggestion, which was to change this to "const auto* iter = blah". 

I wonder if `auto iter` deduces `auto` to be `std::vector<Tensor>::const_iterator`. Either way, I think we should change the code back to
```
for (auto iter = tensor_2D.begin() ...)
```
because the const correctness is not worth complicating the code with the additional `std::vector<Tensor>::const_iterator iter;` line